src/table.rs

```diff
@@ -1,36 +1,28 @@
-use std::sync::atomic::{AtomicU32, Ordering};
-
-type Collation = Option<String>;
-
-pub trait Next {
-    fn new() -> Self;
-    fn next(self) -> Option<Self>
-    where
-        Self: Sized;
-}
-
-impl Next for Collation {
-    fn new() -> Self {
-        Collation::None
-    }
-
-    fn next(self) -> Option<Self> {
-        match self {
-            None => Some(Some("utf8mb4_unicode_ci".to_owned())),
-            Some(x) => match x.as_str() {
-                "utf8mb4_unicode_ci" => Some(Some("utf8mb4_general_ci".to_owned())),
-                "utf8mb4_general_ci" => Some(Some("utf8mb4_bin".to_owned())),
-                "utf8mb4_bin" => None,
-                _ => panic!("unexpected collation"),
-            },
+use async_stream::stream;
+use futures_core::stream::Stream;
+use futures_util::pin_mut;
+use futures_util::stream::StreamExt;
+
+struct Collation;
+impl Collation {
+    fn stream() -> impl Stream<Item = Option<String>> {
+        stream! {
+            for c in vec![
+                None,
+                Some("utf8mb4_unicode_ci".to_owned()),
+                Some("utf8mb4_general_ci".to_owned()),
+                Some("utf8mb4_bin".to_owned()),
+            ]{
+                yield c;
+            }
```

Comment:
Can these stream implementations (except the last cartesian product one) be simplified as static slices?

Reply:
Absolutely they can. I was also thinking whether it's simpler when I rewrote all of them using stream. I decided to do so to avoid mixing the usage of different patterns. I think it would reduce some cognitive burden, at the cost of a bit complicated code.

Reply:
Emmmm... I'm thinking maybe it's simpler if the choices of each part is listed with a simple slice,  and the stream with yield syntax just provides a convenient way to produce the cartesian products of each parts.

Reply:
For parts that are simple enough (like a list), I think either way is easy to understand.
Another problem is that if any part uses the async stream, then parts containing it had better also use it, to `.await`. And we can't predict future usages?

Reply:
An iterator based on another async stream had better to be an async stream too, but I think there isn't any problem to use a plain iterator in an async stream.

Actually I think I'm ok with it if you like this way. It's just not the same way of "simple" as in my mind 😂

Reply:
All right let's leave it as is for now, and modify it when necessary.